File: Python/most_stones_removed_with_same_row_or_column.py
def removeStones(stones: list) -> int:
    def find(count: int, index: int) -> int:
        if parent.get(index, -1) == -1:
            parent[index] = index
            count += 1
        if index != parent.get(index):
            parent[index], count = find(count, parent[index])
        return parent[index], count

    def union(count: int, index1: int, index2: int):
        rootX, count = find(count, index1)
        rootY, count = find(count, index2)
        if rootX == rootY:
            return count
        parent[rootX] = rootY
        count -= 1
        return count

    if len(stones) <= 1:
        return 0

    parent = {}
    
    count = 0
    for stone in stones:
        count = union(count, stone[0] + 10001, stone[1])
    return len(stones) - count


if __name__ == "__main__":
    print(removeStones([[0,0],[0,2],[1,1],[2,0],[2,2]]))

File: src/tmp/Java/src/com/dbc/algorithm/UnionFindSets/MostStonesRemovedWithSameRowOrColumn.java
package com.dbc.algorithm.UnionFindSets;

import java.util.HashMap;
import java.util.Map;

public class MostStonesRemovedWithSameRowOrColumn {
    private int count = 0;
    private Map<Integer, Integer> parent = new HashMap<>();

    public int removeStones(int[][] stones) {
        for (int[] stone : stones) {
            union(stone[0] + 10001, stone[1]);
        }
        return stones.length - this.count;
    }

    public int find(int index) {
        if (!this.parent.containsKey(index)) {
            this.parent.put(index, index);
            this.count++;
        }
        if (index != this.parent.get(index)) {
            this.parent.put(index, find(this.parent.get(index)));
        }
        return this.parent.get(index);
    }

    public void union(int index1, int index2) {
        int rootX = find(index1);
        int rootY = find(index2);
        if (rootX == rootY) {
            return;
        }
        this.parent.put(rootX, rootY);
        count--;
    }
}
